### PR TITLE
Add threading test for Regex

### PR DESCRIPTION
**Pull request by stephentoub**

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs [hunk 1]
```diff
@@ -1242,5 +1244,51 @@ public void Synchronized()
 
             AssertExtensions.Throws<ArgumentNullException>("inner", () => System.Text.RegularExpressions.Match.Synchronized(null));
         }
+
+        public static IEnumerable<object[]> UseRegexConcurrently_ThreadSafe_Success_MemberData()
+        {
+            foreach (TimeSpan timeout in new[] { Timeout.InfiniteTimeSpan, TimeSpan.FromMinutes(1) })
```
Comment (danmoseley) [inline src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs:1250]: curious why varying this parameter is interesting?

**Review by danmoseley — APPROVED** [review]

File: src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs:1250]: It's the exact state I called out on Teams as being buggy in the DFA code? :)

File: src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs [hunk 2]
```diff
@@ -1242,5 +1244,51 @@ public void Synchronized()
...
+                                     m = m.NextMatch();
+                                     Assert.NotNull(m);
+                                     Assert.True(m.Success);
+                                     Assert.Equal("nmlkjihgfedcbax", m.Value);
+
+                                     m = m.NextMatch();
+                                     Assert.NotNull(m);
+                                     Assert.False(m.Success);
+                                 }
+                             }, CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default)).ToArray());
```
Comment (danmoseley) [inline src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs:1290]: I note Olli's hint that starting the regexes at the same moment is most likely to stress the DFA engine. But I wonder whether there might also be value in a variation of this that just plowed through the same 1000's of regex executions you're doing here, but without attempting to start them together. You'd get every possible variation of overlap between them.

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs:1290]: I'm not clear on what you're suggesting. Each thread here is doing 20 matches on the same Regex.  You're suggesting bumping that up to 1000?

File: src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs (hunk 1 above)
Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs:1250]: ah, I didn't see that comment.

File: src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs (hunk 2 above)
Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs:1290]: eh, it didn't make sense.

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs:1290]: Hah, ok :)

**Pull request merged**